packages/gatsby-source-graphql/src/batching/merge-queries.ts

```diff
@@ -0,0 +1,410 @@
+      const { selections } = def.selectionSet
+      return {
+        ...def,
+        selectionSet: {
+          ...def.selectionSet,
+          selections: aliasFieldsInSelection(prefix, selections, doc),
+        },
+      }
+    },
+  }
+  return visit(doc, transformer, { [Kind.DOCUMENT]: [`definitions`] } as any)
+}
+
+/**
+ * Add aliases to fields of the selection, including top-level fields of inline fragments.
+ * Fragment spreads are converted to inline fragments and their top-level fields are also aliased.
+ *
+ * Note that this method is shallow. It adds aliases only to the top-level fields and doesn't
+ * descend to field sub-selections.
+ *
+ * For example, transforms:
+ *   {
+ *     foo
+ *     ... on Query { foo }
+ *     ...FragmentWithBarField
+ *   }
+ * To:
+ *   {
+ *     gatsby1_foo: foo
+ *     ... on Query { gatsby1_foo: foo }
+ *     ... on Query { gatsby1_bar: bar }
+ *   }
+ */
+function aliasFieldsInSelection(
+  prefix: string,
+  selections: ReadonlyArray<SelectionNode>,
+  document: DocumentNode
+): SelectionNode[] {
+  return _.flatMap(selections, (selection: SelectionNode): SelectionNode[] => {
+    switch (selection.kind) {
+      case Kind.INLINE_FRAGMENT:
+        return [aliasFieldsInInlineFragment(prefix, selection, document)]
+
+      case Kind.FRAGMENT_SPREAD: {
+        const inlineFragment = inlineFragmentSpread(selection, document)
+        return [
+          addSkipDirective(selection),
+          aliasFieldsInInlineFragment(prefix, inlineFragment, document),
+          // Keep original spread in selection with @skip(if: true)
+          // otherwise if this was a single fragment usage the query will fail validation
+        ]
+      }
+
+      case Kind.FIELD:
+      default:
+        return [aliasField(selection, prefix)]
+    }
+  })
+}
+
+interface INodeWithDirectives {
+  readonly directives?: ReadonlyArray<DirectiveNode>
+}
+
+function addSkipDirective<T extends INodeWithDirectives>(node: T): T {
+  const skipDirective: DirectiveNode = {
+    kind: Kind.DIRECTIVE,
+    name: { kind: Kind.NAME, value: `skip` },
+    arguments: [
+      {
+        kind: Kind.ARGUMENT,
+        name: { kind: Kind.NAME, value: `if` },
+        value: { kind: Kind.BOOLEAN, value: true },
+      },
+    ],
+  }
+  return {
+    ...node,
+    directives: [skipDirective],
+  }
+}
+
+/**
+ * Add aliases to top-level fields of the inline fragment.
+ * Returns new inline fragment node.
+ *
+ * For Example, transforms:
+ *   ... on Query { foo, ... on Query { bar: foo } }
+ * To
+ *   ... on Query { gatsby1_foo: foo, ... on Query { gatsby1_bar: foo } }
+ */
+function aliasFieldsInInlineFragment(
+  prefix: string,
+  fragment: InlineFragmentNode,
+  document: DocumentNode
+): InlineFragmentNode {
+  const { selections } = fragment.selectionSet
+  return {
+    ...fragment,
+    selectionSet: {
+      ...fragment.selectionSet,
+      selections: aliasFieldsInSelection(prefix, selections, document),
+    },
+  }
+}
+
+/**
```

Comment:
One alternative would be to prefix by fragment also (like you do with query), but it might be harder to deconstruct later.

Reply:
Yeah, that's a part of the tradeoff - prefixing all fragments or trying to de-duplicate and re-use fragments (without variables). We can reconsider if this proves to be a wrong decision, but for now, I'd prefer to keep it as is.